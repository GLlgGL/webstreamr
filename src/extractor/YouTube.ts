import { Context, Format, Meta, UrlResult } from '../types';
import { Extractor } from './Extractor';

export class YouTube extends Extractor {
  public readonly id = 'youtube';

  public readonly label = 'YouTube';

  public override readonly ttl: number = 21600000; // 6h

  public supports(_ctx: Context, url: URL): boolean {
    return null !== url.host.match(/youtube/) && url.searchParams.has('v');
  }

  protected async extractInternal(ctx: Context, url: URL, meta: Meta): Promise<UrlResult[]> {
    const html = await this.fetcher.text(ctx, url);

    const titleMatch = html.match(/"title":{"runs":\[{"text":"(.*?)"/) as string[];

    return [
      {
        url,
        format: Format.unknown,
        ytId: url.searchParams.get('v') as string,
        label: this.label,
        sourceId: `${this.id}_${meta.countryCodes?.join('_')}`,
        ttl: this.ttl,
        meta: {
          ...meta,
          title: titleMatch[1] as string,
        },
      },
    ];
  };
}
