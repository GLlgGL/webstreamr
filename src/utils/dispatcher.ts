import { socksDispatcher } from 'fetch-socks';
import { Agent, Dispatcher, interceptors, ProxyAgent } from 'undici';

const createBasicDispatcher = (): Dispatcher => {
  if (process.env['ALL_PROXY']) {
    const proxyUrl = new URL(process.env['ALL_PROXY']);
    if (proxyUrl.protocol === 'socks5:') {
      return socksDispatcher({ type: 5, host: proxyUrl.hostname, port: parseInt(proxyUrl.port) }, { allowH2: true });
    } else {
      return new ProxyAgent({ uri: proxyUrl.href, allowH2: true });
    }
  } else {
    return new Agent({ allowH2: true });
  }
};

export const createDispatcher = (): Dispatcher => {
  return createBasicDispatcher().compose(
    interceptors.dns(),
    interceptors.retry({ maxRetries: 3 }),
  );
};
