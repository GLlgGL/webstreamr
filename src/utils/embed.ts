import { unpack } from 'unpacker';

export const unpackEval = (html: string): string => {
  const evalMatch = html.match(/eval\(function\(p,a,c,k,e,d\).*\)\)/);
  if (!evalMatch) {
    throw new Error(`No p.a.c.k.e.d string found`);
  }

  return unpack(evalMatch[0]);
};

export const extractUrlFromPacked = (html: string, linkRegExps: RegExp[]): URL => {
  const unpacked = unpackEval(html);

  for (const linkRegexp of linkRegExps) {
    const linkMatch = unpacked.match(linkRegexp);
    if (linkMatch && linkMatch[1]) {
      return new URL(`https://${linkMatch[1].replace(/^(https:)?\/\//, '')}`);
    }
  }

  throw new Error(`Could not find a stream link in embed`);
};
