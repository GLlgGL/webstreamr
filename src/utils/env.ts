import * as os from 'node:os';
import { Request } from 'express';

export const envGet = (name: string): string | undefined => process.env[name];

export const envGetRequired = (name: string): string => {
  const value = envGet(name);
  if (!value) {
    throw new Error(`Environment variable "${name}" is not configured.`);
  }

  return value;
};

export const envGetAppId = (): string => process.env['MANIFEST_ID'] || 'webstreamr';

export const envGetAppName = (): string => process.env['MANIFEST_NAME'] || 'WebStreamr';

export const envIsProd = (): boolean => process.env['NODE_ENV'] === 'production';

export const isElfHostedInstance = (req: Request): boolean => req.host.endsWith('elfhosted.com');

export const getCacheDir = (): string => envGet('CACHE_DIR') ?? os.tmpdir();
