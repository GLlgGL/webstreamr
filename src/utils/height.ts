import { Context } from '../types';
import { CustomRequestInit, Fetcher } from './Fetcher';

export const guessHeightFromPlaylist = async (ctx: Context, fetcher: Fetcher, url: URL, init?: CustomRequestInit): Promise<number | undefined> => {
  const m3u8Data = await fetcher.text(ctx, url, init);

  const heights = Array.from(m3u8Data.matchAll(/\d+x(\d+)|(\d+)p/g))
    .map(heightMatch => heightMatch[1] ?? heightMatch[2])
    .filter(height => height !== undefined)
    .map(height => parseInt(height));

  return heights.length ? Math.max(...heights) : /* istanbul ignore next */ undefined;
};
