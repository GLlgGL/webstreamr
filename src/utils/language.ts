import { CountryCode } from '../types';

const countryCodeMap: Record<CountryCode, { language: string; flag: string; iso639: string | undefined }> = {
  multi: { language: 'Multi', flag: '🌐', iso639: undefined },
  de: { language: 'German', flag: '🇩🇪', iso639: 'ger' },
  en: { language: 'English', flag: '🇺🇸', iso639: 'eng' },
  es: { language: 'Castilian Spanish', flag: '🇪🇸', iso639: 'spa' },
  fr: { language: 'French', flag: '🇫🇷', iso639: 'fra' },
  hi: { language: 'Hindi', flag: '🇮🇳', iso639: 'hin' },
  it: { language: 'Italian', flag: '🇮🇹', iso639: 'ita' },
  ja: { language: 'Japanese', flag: '🇯🇵', iso639: 'jpn' },
  ko: { language: 'Korean', flag: '🇰🇷', iso639: 'kor' },
  mx: { language: 'Latin American Spanish', flag: '🇲🇽', iso639: 'spa' },
};

export const languageFromCountryCode = (countryCode: CountryCode) => {
  return countryCodeMap[countryCode].language;
};

export const flagFromCountryCode = (countryCode: CountryCode) => {
  return countryCodeMap[countryCode].flag;
};

export const iso639FromCountryCode = (countryCode: CountryCode) => {
  return countryCodeMap[countryCode].iso639;
};

export const findCountryCodes = (value: string): CountryCode[] => {
  const countryCodes: CountryCode[] = [];

  for (const countryCode in countryCodeMap) {
    if (value.includes(countryCodeMap[countryCode as CountryCode]['language'])) {
      countryCodes.push(countryCode as CountryCode);
    }
  }

  return countryCodes;
};
