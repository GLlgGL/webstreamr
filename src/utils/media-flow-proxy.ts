import { Context } from '../types';
import { Fetcher } from './Fetcher';

interface ExtractResult {
  destination_url: string;
  request_headers: Record<string, string>;
  mediaflow_proxy_url: string;
  query_params: Record<string, string>;
}

export const supportsMediaFlowProxy = (ctx: Context): boolean => !!ctx.config['mediaFlowProxyUrl'];

const buildMediaFlowProxyExtractorUrl = (ctx: Context, host: string, url: URL, headers: Record<string, string>): URL => {
  const mediaFlowProxyUrl = new URL('/extractor/video', ctx.config.mediaFlowProxyUrl);

  mediaFlowProxyUrl.searchParams.append('host', host);
  mediaFlowProxyUrl.searchParams.append('api_password', `${ctx.config.mediaFlowProxyPassword}`);
  mediaFlowProxyUrl.searchParams.append('d', url.href);

  for (const headerKey in headers) {
    mediaFlowProxyUrl.searchParams.set('h_' + headerKey.toLowerCase(), headers[headerKey] as string);
  }

  return mediaFlowProxyUrl;
};

export const buildMediaFlowProxyExtractorRedirectUrl = (ctx: Context, host: string, url: URL, headers: Record<string, string> = {}): URL => {
  const mediaFlowProxyUrl = buildMediaFlowProxyExtractorUrl(ctx, host, url, headers);

  mediaFlowProxyUrl.searchParams.append('redirect_stream', 'true');

  return mediaFlowProxyUrl;
};

export const buildMediaFlowProxyExtractorStreamUrl = async (ctx: Context, fetcher: Fetcher, host: string, url: URL, headers: Record<string, string>): Promise<URL> => {
  const mediaFlowProxyUrl = buildMediaFlowProxyExtractorUrl(ctx, host, url, headers);

  const extractResult: ExtractResult = await fetcher.json(ctx, mediaFlowProxyUrl, { queueLimit: 4, queueTimeout: 10000, timeout: 20000 });

  const streamUrl = new URL(extractResult.mediaflow_proxy_url);

  for (const queryParamsKey in extractResult.query_params) {
    streamUrl.searchParams.append(queryParamsKey, extractResult.query_params[queryParamsKey] as string);
  }
  for (const requestHeadersKey in extractResult.request_headers) {
    streamUrl.searchParams.append(`h_${requestHeadersKey}`, extractResult.request_headers[requestHeadersKey] as string);
  }
  streamUrl.searchParams.append('d', extractResult.destination_url);

  return streamUrl;
};
