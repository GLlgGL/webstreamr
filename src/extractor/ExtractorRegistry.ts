import TTLCache from '@isaacs/ttlcache';
import winston from 'winston';
import { Context, CountryCode, UrlResult } from '../types';
import { isExtractorDisabled } from '../utils';
import { Extractor } from './Extractor';

export class ExtractorRegistry {
  private readonly logger: winston.Logger;
  private readonly extractors: Extractor[];
  private readonly urlResultCache: TTLCache<string, UrlResult[]>;

  public constructor(logger: winston.Logger, extractors: Extractor[]) {
    this.logger = logger;
    this.extractors = extractors;
    this.urlResultCache = new TTLCache({ max: 4096 });
  }

  public async handle(ctx: Context, url: URL, countryCode: CountryCode, title?: string | undefined): Promise<UrlResult[]> {
    const extractor = this.extractors.find(extractor => !isExtractorDisabled(ctx.config, extractor) && extractor.supports(ctx, url));
    if (!extractor) {
      return [];
    }

    const normalizedUrl = extractor.normalize(url);
    const cacheKey = extractor.viaMediaFlowProxy
      ? `${extractor.id}_${normalizedUrl}_${ctx.config.mediaFlowProxyUrl}`
      : `${extractor.id}_${normalizedUrl}`;

    let urlResults = this.urlResultCache.get(cacheKey) ?? [];
    if (this.urlResultCache.has(cacheKey)) {
      return urlResults.map(urlResult => ({ ...urlResult, ttl: this.urlResultCache.getRemainingTTL(cacheKey) }));
    }

    this.logger.info(`Extract stream URL using ${extractor.id} extractor from ${url}`, ctx);

    urlResults = await extractor.extract(ctx, normalizedUrl, countryCode, title);

    if (!urlResults.some(urlResult => urlResult.error) && extractor.ttl) {
      this.urlResultCache.set(cacheKey, urlResults, { ttl: extractor.ttl });
    }

    return urlResults;
  };
}
