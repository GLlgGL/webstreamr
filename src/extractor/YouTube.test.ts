import winston from 'winston';
import { createTestContext } from '../test';
import { FetcherMock } from '../utils';
import { ExtractorRegistry } from './ExtractorRegistry';
import { YouTube } from './YouTube';

const logger = winston.createLogger({ transports: [new winston.transports.Console({ level: 'nope' })] });
const extractorRegistry = new ExtractorRegistry(logger, [new YouTube(new FetcherMock(`${__dirname}/__fixtures__/YouTube`))]);

const ctx = createTestContext();

describe('YouTube', () => {
  test('Solaris', async () => {
    expect(await extractorRegistry.handle(ctx, new URL('https://www.youtube.com/watch?v=Z8ZhQPaw4rE'))).toMatchSnapshot();
  });

  test('unsupported format', async () => {
    expect(await extractorRegistry.handle(ctx, new URL('https://youtu.be/Z8ZhQPaw4rE?feature=shared'))).toMatchSnapshot();
  });
});
