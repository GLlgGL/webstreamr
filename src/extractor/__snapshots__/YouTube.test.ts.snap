// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`YouTube Solaris 1`] = `
[
  {
    "format": "unknown",
    "label": "YouTube",
    "meta": {
      "countryCodes": [],
      "title": "Solaris | SCIENCE FICTION | FULL MOVIE | directed by Tarkovsky",
    },
    "sourceId": "youtube_",
    "ttl": 21600000,
    "url": "https://www.youtube.com/watch?v=Z8ZhQPaw4rE",
    "ytId": "Z8ZhQPaw4rE",
  },
]
`;

exports[`YouTube unsupported format 1`] = `[]`;
